File: src/01-easy/happy-numbers/solution.php
<?php
$n = intval(trim(fgets(STDIN)));
for ($i = 0; $i < $n; $i++) {
    $s = trim(fgets(STDIN));
    $x = array_sum(array_map(fn($c) => ($c) ** 2, str_split($s)));
    $seen = [];
    while ($x != 1 && !isset($seen[$x])) { $seen[$x] = 1; $t = 0; while ($x > 0) { $d = $x % 10; $t += $d*$d; $x = intdiv($x, 10); } $x = $t; }
    echo "$s " . ($x == 1 ? ":)" : ":(") . "\n";
}

File: src/01-easy/happy-numbers/solution.pl
$| = 1;
my $n = <STDIN>; chomp $n;
for (1..$n) {
    chomp(my $s = <STDIN>);
    my $x = 0; $x += $_**2 for split //, $s;
    my %seen;
    while ($x != 1 && !$seen{$x}++) {
        my $t = 0; $t += $_**2 for split //, $x; $x = $t;
    }
    print "$s " . ($x == 1 ? ":)" : ":(") . "\n";
}
